### PR TITLE
Changed in nnx.jit in/out sharding default values to UNSPECIFIED as in jax.jit

### LINKED ISSUE #4791: nnx.jit doesn't follow jax.sharding.use_mesh
```py
mesh = jax.make_mesh((2, 2), ('x', 'y'))
pspec = jax.sharding.PartitionSpec("x", "y")
f1 = jax.jit(lambda: jax.lax.with_sharding_constraint(jnp.ones((4, 4)), pspec))
f2 = nnx.jit(lambda: jax.lax.with_sharding_constraint(jnp.ones((4, 4)), pspec))

with mesh:
  print(f1().sharding.spec, f2().sharding.spec)

with jax.sharding.use_mesh(mesh):
  print(f1().sharding.spec, f2().sharding.spec)
```

Output:

```
PartitionSpec('x', 'y') PartitionSpec('x', 'y')
PartitionSpec('x', 'y') PartitionSpec()
```

My understanding is `use_mesh` is the new mesh context API which will replace the old `with mesh` syntax. @yashk2810 to confirm.

### PR DESCRIPTION
**Pull request by vfdev-5**
# What does this PR do?

Fixes #4791 

- Changed in nnx.jit in/out sharding default values to UNSPECIFIED as in jax.jit

## Checklist
- [ ] This PR fixes a minor issue (e.g.: typo or small bug) or improves the docs (you can dismiss the other
      checks if that's the case).
- [ ] This change is discussed in a Github issue/
      [discussion](https://github.com/google/flax/discussions) (please add a
      link).
- [ ] The documentation and docstrings adhere to the
      [documentation guidelines](https://github.com/google/flax/blob/main/docs/README.md#how-to-write-code-documentation).
- [ ] This change includes necessary high-coverage tests.
      (No quality testing = no merge!)

### CODE REVIEW

File: tests/nnx/transforms_test.py [hunk 1]
```diff
@@ -497,10 +524,10 @@ def test_basic_shardmap_variables(self):
...
 
     @nnx.shard_map(mesh=mesh, in_specs=(P(None, 'a'), P(), P()), out_specs=None)
-    def f(w, b, count):
-      count += 1
-      self.assertEqual(w.shape, (w.shape[0], w.shape[1] // n_devices))
-      self.assertEqual(b.shape, (32,))
+    def f(w_, b_, count_):
+      count_ += 1
+      self.assertEqual(w_.shape, (w.shape[0], w.shape[1] // n_devices))
+      self.assertEqual(b_.shape, (32,))
```
Comment (vfdev-5) [inline tests/nnx/transforms_test.py:530]: Drive by test fix as it wont pass with n_devices > 1

File: tests/nnx/transforms_test.py [hunk 2]
```diff
@@ -55,6 +55,10 @@ def __getattr__(self, key): ...
 
 
 class TestJIT(absltest.TestCase):
+  @classmethod
+  def setUpClass(cls):
+    jax.config.update('jax_num_cpu_devices', 4)
```
Comment (vfdev-5) [inline tests/nnx/transforms_test.py:60]: I'm not totally sure how safe this line can be in the context whether this test class could be called on a large infrastructure with multiple devices.

File: flax/nnx/transforms/compilation.py [hunk 3]
```diff
@@ -22,6 +22,7 @@
 import jax.experimental
 import jax.experimental.shard_map
 from jax.sharding import AbstractMesh, Mesh, PartitionSpec
+from jax._src import sharding_impls
```
Comment (yashk2810) [inline flax/nnx/transforms/compilation.py:25]: Please do not use `_src` imports from JAX.

File: tests/nnx/transforms_test.py (hunk 2 above)
Comment (cgarciae) [reply tests/nnx/transforms_test.py:60]: maybe you can add the tear down method and set it back to the original value?
Also, not sure how many cores Github CI machines have.

Comment (vfdev-5) [reply tests/nnx/transforms_test.py:60]: By default, they are:

Virtual Machine | Processor (CPU) | Memory (RAM) | Storage (SSD) | Architecture | Workflow label
-- | -- | -- | -- | -- | --
Linux | 4 | 16 GB | 14 GB | x64 | ubuntu-latest, ubuntu-24.04, ubuntu-22.04


https://docs.github.com/en/actions/how-tos/using-github-hosted-runners/using-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories

**Comment by vfdev-5** [comment]
I close this PR as the issue related to nnx.jit is solved on jax side by this PR: https://github.com/jax-ml/jax/pull/29960
I'll send fixes for the test suite in another PR.
